File: basic-proxy-impl/src/test/java/com/test/util/SmartInvocationHandler.java
package com.test.util;

import org.objectweb.asm.Type;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class SmartInvocationHandler implements InvocationHandler {

    private final Set<Method> calledMethods = new HashSet<>();

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) {
        calledMethods.add(method);

        Type returnType = Type.getReturnType(method);
        if (returnType == Type.BOOLEAN_TYPE) {
            return Boolean.FALSE;
        }
        if (returnType == Type.CHAR_TYPE) {
            return Character.MIN_VALUE;
        }
        if (returnType == Type.SHORT_TYPE) {
            return (short) 0;
        }
        if (returnType == Type.BYTE_TYPE) {
            return (byte) 0;
        }
        if (returnType == Type.INT_TYPE) {
            return 0;
        }
        if (returnType == Type.FLOAT_TYPE) {
            return 0.0f;
        }
        if (returnType == Type.LONG_TYPE) {
            return 0L;
        }
        if (returnType == Type.DOUBLE_TYPE) {
            return 0.0d;
        }
        return null;
    }

    public Set<Method> getCalledMethods() {
        return Collections.unmodifiableSet(calledMethods);
    }
}

File: basic-proxy-kotlin-impl/src/test/kotlin/com/test/util/SmartInvocationHandler.kt
package com.test.util

import org.objectweb.asm.Type
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

class SmartInvocationHandler : InvocationHandler {

    private val calledMethods = mutableSetOf<Method>()

    override fun invoke(
        proxy: Any,
        method: Method,
        args: Array<out Any?>?
    ): Any? {
        calledMethods.add(method)

        return when (Type.getReturnType(method)) {
            Type.BOOLEAN_TYPE -> false
            Type.CHAR_TYPE -> Character.MIN_VALUE
            Type.SHORT_TYPE -> 0.toShort()
            Type.BYTE_TYPE -> 0.toByte()
            Type.INT_TYPE -> 0
            Type.FLOAT_TYPE -> 0.toFloat()
            Type.LONG_TYPE -> 0.toLong()
            Type.DOUBLE_TYPE -> 0.0
            else -> {
                null
            }
        }
    }

    fun getCalledMethods(): Set<Method> {
        return calledMethods.toSet()
    }
}
